examples/scripts/denoise_cplx_tv_pdhg.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# This file is part of the SCICO package. Details of the copyright
+# and user license can be found in the 'LICENSE.txt' file distributed
+# with the package.
+
+r"""
+Complex Total Variation Denoising (ADMM)
+========================================
+
+This example demonstrates solution of a problem of the form
+
+  $$\argmin_{\mathbf{x}} \; f(\mathbf{x}) + g(C(\mathbf{x})) \;,$$
+
+where $C$ is a nonlinear operator, via non-linear PDHG
+:cite:`valkonen-2014-primal`. The example problem represents total
+variation (TV) denoising applied to a complex image with piece-wise
+smooth magnitude and non-smooth phase. The appropriate TV denoising
+formulation for this problem is
+
+  $$\mathrm{argmin}_{\mathbf{x}} \; (1/2) \| \mathbf{y} - \mathbf{x}
+  \|_2^2 + \lambda \| C(\mathbf{x}) \|_{2,1} \;,$$
+
+where $\mathbf{y}$ is the measurement, $\|\cdot\|_{2,1}$ is the
+$\ell_{2,1}$ mixed norm, and $C$ is a non-linear operator that applies a
+linear difference operator to the magnitude of a complex array. The
+standard TV solution, which is also computed for comparison purposes,
+gives very poor results since the difference is applied independently to
+real and imaginary components of the complex image.
+"""
+
+import jax
+
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+from xdesign import SiemensStar, discrete_phantom
+
+import scico.numpy as snp
+import scico.random
+from scico import functional, linop, loss, metric, operator, plot
+from scico.examples import phase_diff
+from scico.optimize import PDHG
+from scico.util import device_info
+
+"""
+Create a ground truth image.
+"""
+N = 256  # image size
+phantom = SiemensStar(16)
+x_mag = snp.pad(discrete_phantom(phantom, N - 16), 8) + 1.0
+x_mag /= x_mag.max()
+# Create reference image with structured magnitude and random phase
+x_gt = x_mag * snp.exp(-1j * scico.random.randn(x_mag.shape, seed=0)[0])
+x_gt = jax.device_put(x_gt)  # convert to jax type, push to device
+
+
+"""
+Add noise to create a noisy test image.
+"""
+σ = 0.25  # noise standard deviation
+noise, key = scico.random.randn(x_gt.shape, seed=1, dtype=snp.complex64)
+y = x_gt + σ * noise
+
+
+"""
+Denoise with standard total variation.
+"""
+λ_tv = 6e-2
```

Comment:
The result for standard TV looks undersmoothed. Can this be increased? Or is the problem that somehow the Re and Im part _are_ smooth, but the mag/phase representation still looks noisy?

Reply:
The magnitude is smooth and the phase is noise. Pushing up the regularization improves the magnitude, but it smooths the phase more, which drops the SNR. I didn't do an automated parameter search, but I hand tuned to what I believe is pretty close to the minimum MSE setting.